File: routes/raid.js
module.exports = function (fastify, opts, done) {
	const Router = fastify;

	Router.get("/", async (req, res) => {
		const region = req.query.region || "global";
		if (!["global", "japan"].includes(region)) {
			return res.badRequest("Invalid region");
		}

		const raidData = await ba.Query.collection(`${region}.RaidData`).find({}).toArray();
		if (!raidData || raidData.length === 0) {
			return res.notFound();
		}

		const stages = {
			current: [],
			upcoming: [],
			ended: []
		};

		for (const raid of raidData) {
			const startAt = raid.startAt;
			const endAt = raid.endAt;
			const now = new Date();

			if (now > startAt && now < endAt) {
				stages.current.push({
					seasonId: raid.id,
					bossName: raid.boss,
					startAt: raid.startAt,
					settleAt: raid.settleAt,
					endAt: raid.endAt
				});
			}
			else if (now < startAt) {
				stages.upcoming.push({
					seasonId: raid.id,
					bossName: raid.boss,
					startAt: raid.startAt,
					settleAt: raid.settleAt,
					endAt: raid.endAt
				});
			}
			else if (now > endAt) {
				stages.ended.push({
					seasonId: raid.id,
					bossName: raid.boss,
					startAt: raid.startAt,
					settleAt: raid.settleAt,
					endAt: raid.endAt
				});
			}
		}

		return res.send(stages);
	});

	done();
};

File: internal/api/raid.go
package api

import (
	"context"
	"log"
	"net/http"
	"sort"
	"strings"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"

	"go-archive/internal/http/respond"
)

const queryTimeout = 5 * time.Second

// RaidSeason mirrors the MongoDB raid document structure.
type RaidSeason struct {
	ID       int    `bson:"id" json:"id"`
	Boss     string `bson:"boss" json:"boss"`
	StartAt  int64  `bson:"startAt" json:"startAt"`
	EndAt    int64  `bson:"endAt" json:"endAt"`
	SettleAt int64  `bson:"settleAt,omitempty" json:"settleAt,omitempty"`
}

// RaidResponse groups raid seasons by lifecycle.
type RaidResponse struct {
	Current  []RaidSeason `json:"current"`
	Upcoming []RaidSeason `json:"upcoming"`
	Ended    []RaidSeason `json:"ended"`
}

// NewRaidHandler returns an HTTP handler backed by MongoDB raid collections.
func NewRaidHandler(client *mongo.Client, database string) http.HandlerFunc {
	if client == nil {
		return func(w http.ResponseWriter, _ *http.Request) {
			respond.Error(w, http.StatusInternalServerError, "database client not configured")
		}
	}

	empty := RaidResponse{
		Current:  []RaidSeason{},
		Upcoming: []RaidSeason{},
		Ended:    []RaidSeason{},
	}

	collectionByRegion := map[string]string{
		"global": "global.RaidData",
		"japan":  "japan.RaidData",
	}

	return func(w http.ResponseWriter, r *http.Request) {
		region := strings.ToLower(strings.TrimSpace(r.URL.Query().Get("region")))
		if region == "" {
			region = "global"
		}
		collectionName, ok := collectionByRegion[region]
		if !ok {
			respond.Error(w, http.StatusBadRequest, "invalid region; use global or japan")
			return
		}

		ctx, cancel := context.WithTimeout(r.Context(), queryTimeout)
		defer cancel()

		collection := client.Database(database).Collection(collectionName)
		cursor, err := collection.Find(ctx, bson.M{})
		if err != nil {
			log.Printf("raid find error (%s): %v", region, err)
			respond.Error(w, http.StatusInternalServerError, "failed to load raid data")
			return
		}
		defer cursor.Close(ctx)

		var records []RaidSeason
		if err := cursor.All(ctx, &records); err != nil {
			log.Printf("raid decode error (%s): %v", region, err)
			respond.Error(w, http.StatusInternalServerError, "failed to decode raid data")
			return
		}

		if len(records) == 0 {
			respond.JSON(w, http.StatusOK, empty)
			return
		}

		now := time.Now().UnixMilli()
		response := RaidResponse{
			Current:  make([]RaidSeason, 0),
			Upcoming: make([]RaidSeason, 0),
			Ended:    make([]RaidSeason, 0),
		}

		for _, raid := range records {
			switch {
			case raid.StartAt <= now && raid.EndAt >= now:
				response.Current = append(response.Current, raid)
			case raid.StartAt > now:
				response.Upcoming = append(response.Upcoming, raid)
			default:
				response.Ended = append(response.Ended, raid)
			}
		}

		sort.Slice(response.Current, func(i, j int) bool {
			return response.Current[i].StartAt < response.Current[j].StartAt
		})
		sort.Slice(response.Upcoming, func(i, j int) bool {
			return response.Upcoming[i].StartAt < response.Upcoming[j].StartAt
		})
		sort.Slice(response.Ended, func(i, j int) bool {
			return response.Ended[i].StartAt > response.Ended[j].StartAt
		})

		respond.JSON(w, http.StatusOK, response)
	}
}
